docs/adr/0008-freeing-objects-and-resources.md

```diff
@@ -0,0 +1,146 @@
+
+UniFFI needs to ensure that when the generated objects are destroyed, the reference on the other side of the FFI is released.
+It is acceptable for this release to be delayed and/or non-deterministic as long as it eventually happens.
+In particular, it is acceptable for it to be handled by garbage collection.
+
+## Closing Resources
+
+When an object holds a significant resource (eg, some scarce operating-system owned resource), it may not be acceptable to couple the resource to the lifetime of the foreign object.
+In this case, it MUST be to possible close the resource while the foreign object is still alive and before its reference is released.
+
+These same considerations apply in both directions across the FFI.
+If a garbage-collected foreign object stores a resource, then there MUST be a way to close the resource before the garbage collection happens.
+If a Rust object holds a resource, there MUST be a way to close that resource even if there may be other `Arc<>` references.
+
+## Object containers
+
+The discussion above has focused on individual objects, but in any non-trivial application many objects interact.
+For example, consider a simple `Enum` or `dictionary` - in the usual case, these objects hold references to a non-resource holding objects.
+Given the discussion above, it's acceptable for these references to be released using the standard memory management of the foreign language.
+
+However, the `Enum`, `dictionary`, `Vec`, `HashMap`, etc. may contain an object that itself contains a resource.
+While the component author is able to offer methods to close the resource on the objects themselves, they are not able to introduce methods to all the container objects which might hold the object.
+So the question here is whether UniFFI itself should try and offer such capabilities.
+
+## Foreign "sugar" for closing resources.
+
+Objects which hold significant resources offer the capability to closing those resources even while the object remains alive.
+For example, most foreign languages which expose a "file" object tend to provide a `close()` method that ensures the handle is released to the operating system.
+
+For almost all languages it is possible use something like a `try/finally` pattern to ensure the object is closed as soon as practical.
+This pattern is so common that some languages offer syntactic sugar for this express purpose.
+For example, Python offers a "context manager", Kotlin offers the `use` extension for objects that implement `Closable`, etc.
+
+Ideally, UniFFI would offer a way so consumers can opt-in to implementing this capability.
+This must be opt-in because that opt-in process must tell UniFFI exactly how the resource should be closed - while it will be very common for this to be via a `close()` method, UniFFI can't assume this (eg, a Python context manager doesn't insist on any particular technique)
+
+## Decision Drivers
```

Comment:
Form nit: what are the things that make a good option?

e.g. Simplicity? Non-leaky abstractions? Performance?

Reply:
Good point, I added a bullet point and reworked another one to try to address this.